### PR TITLE
skylarkstruct: implement to_proto

### PR DESCRIPTION
**Pull request by alandonovan**

### CODE REVIEW

**Review by jayconrod — APPROVED** [review]

**Comment by mats23177** [comment]
some people approved

**Comment by alandonovan** [comment]
Care to elaborate?

File: skylarkstruct/struct.go [hunk 1]
```diff
@@ -223,7 +217,11 @@ func (s *Struct) Attr(name string) (skylark.Value, error) {
 		return s.entries[i].value, nil
 	}
 
-	// TODO(adonovan): to_{json,proto} are deprecated (b/36412967).
+	// TODO(adonovan): there may be a nice feature for core
+	// skylark.Value here, especially for JSON, but the current
+	// features are incomplete and underspecified.
+	//
+	// to_{json,proto} are deprecated, appropriately; see Google issue b/36412967.
```
Comment (abergmeier) [inline skylarkstruct/struct.go:224]: So 1. we outside of Google can not see any of your internal issues. Referring to them is not the nicest thing to do.
2. This is the first reference I found for `to_`* deprecation. While I can think of a few reasons why - it is unclear what the actual reasons are. Could you please create a public issue for that topic somewhere.
3. Do you already have a Skylark (still waiting on the rename to happen) implementation for `json.dumps`?

Summoning @dslomov

**Comment by alandonovan** [comment]
> we outside of Google can not see any of your internal issues. Referring to them is not the nicest thing to do.

I appreciate that, but unfortunately some of these discussions are initiated by users of Blaze on Google's internal issue tracker, and it would take too much effort to sanitize and re-report every appropriate Blaze issue as an upstream Bazel issue.

> 2. This is the first reference I found for to_* deprecation. While I can think of a few reasons why - it is unclear what the actual reasons are. Could you please create a public issue for that topic somewhere.

There is no official plan to remove them AFAICT.  "Deprecation" usually means little more than "this function is best avoided because it is broken by design and/or poorly maintained."

Here are my personal reasons for disliking the to_json and to_proto methods:
1. Having them be methods of struct creates questions of how to resolve ambiguity in the namespace: what happens if you create a struct with a field named to_proto?
2. the JSON and proto conversions are poorly specified.  There's no reason why such conversions shouldn't be accessible to Skylark programs, but there's no reason struct, but not list or dict, is specially deserving of such methods.
3. to_proto creates an undesirable dependency upon the protocol buffer machinery, which is otherwise not a necessary part of a Skylark interpreter.  Defining it an extension avoids that.  (to_json is similar, but a JSON encoder is much easier to write by hand.)


> 3. Do you already have a Skylark (still waiting on the rename to happen) implementation for json.dumps?

I don't.  What is json.dumps?

[UPDATE: Ah, it's a Python package for printing values as JSON.  That's exactly what I had in mind, but I haven't implemented it.]

("Summoning"... I like that.)

**Comment by abergmeier** [comment]
So focussing on `to_proto`, is there any pure-Skylark (what about the rebranding?) implementation yet?
If not I would like to take a shot at that (if you would accept the contribution). Using the same limitations (16 levels of depth), it should be feasible.

**Comment by alandonovan** [comment]
On 12 March 2018 at 13:14, Andreas Bergmeier <notifications@github.com>
wrote:

> So focussing on to_proto, is there any pure-Skylark (what about the
> rebranding?) implementation yet?
> If not I would like to take a shot at that (if you would accept the
> contribution). Using the same limitations (16 levels of depth), it should
> be feasible.
>

There is not.  I would accept the contribution as an extension package.  It
would be best to circulate a sketch of the API and the proposed mapping
from Skylark values to strings before you do significant work.  Thanks.

Re: the name, they have chosen a new one, but if Bazel team hasn't
announced it yet then it's probably still with the lawyers.

**Comment by abergmeier** [comment]
> It would be best to circulate a sketch of the API

Well API is an overstatement. It probably (for starters) would be a free-standing method named `to_proto` and accept a `struct` same as now and also write the same output.
On the plus side:
- As a drop-in replacement for `struct.to_proto` we could unify the impl between Java and Go.
- More dogfeeding
- Be able to clean up struct long term
- Nobody is happy with the current `struct` behavior.

On the negative side:
- Limit to what can actually be achieved with the Skylark exposed Python mechanics
- Performance (will probably be degraded)

**Pull request merged**